FlintCore/Utils/CompletionRequirement.swift

```diff
@@ -0,0 +1,210 @@
+/// ```
+///
+/// When implementing such a function requiring a completion handler, you return one of two statuses returned by either
+/// the `CompletionRequirement.completed(_ arg: T)` or `CompletionRequirement.willCompleteAsync()`.
+/// The `CompletionRequirement` will take care of calling the completion block as appropriate.
+///
+/// ```
+/// func doSomething(input: Any, completionRequirement: DoSomethingCompletion) -> DoSomethingCompletion.Status {
+///     return completionRequirement.completedSync(false)
+/// }
+///
+/// // or for async completion
+///
+/// func doSomething(input: Any, completionRequirement: DoSomethingCompletion) -> DoSomethingCompletion.Status {
+///     let result = completionRequirement.willCompleteAsync()
+///     DispatchQueue.main.asyncAfter(deadline: .now() + 5) {
+///         result.completed(false)
+///     }
+///     return result
+/// }
+/// ```
+public class CompletionRequirement<T> {
+    public class Status {
+        fileprivate let _value: T?
+        fileprivate var value: T {
+            get {
+                guard let result = _value else {
+                    flintBug("CompletionRequirement status value is nil")
+                }
+                return result
+            }
+        }
+        
+        init() {
+            _value = nil
+        }
+
+        init(result: T) {
+            _value = result
+        }
+        
+        public var isCompletingAsync: Bool { return false }
+    }
+
+    // The type for a status indicating completion will occur later
+    public class DeferredStatus: Status {
+        var owner: CompletionRequirement<T>?
+        
+        init(owner: CompletionRequirement<T>) {
+            self.owner = owner
+            super.init()
+        }
+
+        override public var isCompletingAsync: Bool {
+            return true
+        }
+        
+        public func completed(_ result: T) {
+            guard let owner = owner else {
+                return
+            }
+            owner.completionHandler(result, true)
+        }
+    }
+
+    fileprivate var completionStatus: Status?
+    var completionHandler: ((T, _ completedAsync: Bool) -> Void)!
+
+    public init(completionHandler: @escaping (T, _ completedAsync: Bool) -> Void) {
+        self.completionHandler = completionHandler
+    }
+
+    /// Internal initialiser for proxy subclass, which will set the completion after
+    fileprivate init() {
+    }
+
+    /// Call to verify that the result belongs to this completion instance and there hasn't been a mistake
+    public func verify(_ status: Status) -> Bool {
+        return status === completionStatus
+    }
+    
+    /// Call to indicate that completion will be called later, asynchronously by code that has a reference
+    /// to the deferred status.
+    public func willCompleteAsync() -> DeferredStatus {
+        guard completionStatus == nil else {
+            flintUsageError("Only one of completedSync() or willCompleteLater() can be called")
```

Comment:
Not sure if this type of idea can work or how much of a rework it would be, but what if you have some type of mutable version of this class (like a builder type concept) and once the `completedSync` or `willCompleteLater` is called, it vends the immutable version so this can't happen? Maybe it's not worth the overhead; I'm not sure.

Reply:
That's a nice idea. It's sort of what we're doing with the status though - I think with the builder type you'd still have this same issue - you could call it twice and would need to protect against it?

Reply:
Yeah I guess it wouldn’t change it much because you would still have to call build, so you may still end up changing it twice.